File: go/bestline.go
package main

import "fmt"

type BestLine struct {
	points uint
	best_fit Line
}

func NewBestLine(points []Point) *BestLine {
	if len(points) < 2 {
		zero := Point{0, 0}
		return &BestLine{uint(len(points)), *NewLine(zero, zero)}
	}
	max_i := uint(0)
	max_l := *NewLine(points[0], points[1])
	for i, a := range points {
		for j, b := range points[i+1:] {
			l := *NewLine(a, b)
			intersections := uint(2)
			for _, c := range points[i+j+2:] {
				if l.intersects(c) {
					intersections += 1
				}
			}
			if intersections > max_i {
				max_i = intersections
				max_l = l
			}
		}
	}
	return &BestLine{max_i, max_l}
}

func (bl BestLine) intersections() uint {
	return bl.points
}

func (bl BestLine) String() string {
	return fmt.Sprintf("%v", bl.best_fit)
}

File: dart/lib/bestline.dart
import 'line.dart';
import 'point.dart';

class BestLine {
	int points;
	Line bestFit;

	BestLine(List<Point> points) :
			points = points.length,
			bestFit = Line(Point(0, 0), Point(0, 0)) {
		if (points.length > 1) {
			int maxI = 0;
			var maxL = Line(points[0], points[1]);
			for (int i = 0; i < points.length - 1; i += 1) {
				for (int j = i + 1; j < points.length; j += 1) {
					final a = points[i];
					final b = points[j];
					final l = Line(a, b);
					final intersections = 2 + points.sublist(j + 1)
							.where((p) => l.intersects(p)).length;
					if (intersections > maxI) {
						maxI = intersections;
						maxL = l;
					}
				}
			}
			this.points = maxI;
			bestFit = maxL;
		}
	}

	@override
	String toString() {
		return bestFit.toString();
	}
}
